File: dependency_breaking_techniques/examples/test_override/test_report_generator.cpp
#include<cassert>
#include<vector>

#include "report_generator.h"

using namespace std;

class DummyReportSaver: public IReportSaver {
public:
    void save_report(const Report &) override { }
};

class TestReportGenerator: public ReportGenerator {
public:
    Report report;
protected:
    IReportSaver *make_report_saver() override { return new DummyReportSaver(); }
    void report_data(const Report &report) override { this->report = report; }
};

void test_analyze_data() {
    vector<double> data{1,2,3,4};
    auto report_generator = TestReportGenerator();
    report_generator.analyze_data("test-data", data);
    double mean = report_generator.report.mean;
    double var = report_generator.report.var;
    // verify:
    // mean = 2.5, var = 1.25
    assert(mean == 2.5);
    assert(var == 1.25);
}

int main(int argc, const char **argv) {
    test_analyze_data();
}

File: dependency_breaking_techniques/examples/test_override/test_report_generator.py
from report_generator import ReportGenerator

class DummyReportSaver:
    def save_report(self, *args, **kwargs):
        pass

class TestReportGenerator(ReportGenerator):
    def make_report_saver(self):
        return DummyReportSaver()

    def report_data(self, report):
        self.report = report

def test_analyze_data():
    data = [1,2,3,4]
    # verify:
    #  mean = 2.5, var = 1.25
    report_generator = TestReportGenerator()
    report_generator.analyze_data('test-data', data)
    mean = report_generator.report['mean']
    var = report_generator.report['var']
    assert mean == 2.5
    assert var == 1.25
    
if __name__ == '__main__':
    test_analyze_data()
